### DISCUSSION TITLE
Help with MCMC run

### DISCUSSION THREAD

**Discussion by Nishijannat** [AstroBridge/BDLensing#205] [Q&A] [answered]
[1](https://colab.research.google.com/drive/1cNP2N99l1buQ967ZCSIlKrFrvDcif3OI?usp=sharing),[2](https://colab.research.google.com/drive/1AKaUfEchmcaT7KVbEBLGYW-PnLgyVK00), [3](https://colab.research.google.com/drive/10Yxw4Xkr-87M3WneK1PdAx9v1dq_oari),[4](https://colab.research.google.com/drive/16iDnXLMofPeZ_ry7-FFO98Zy2nucoXfG)

I have trialed mcmc for several times and found some problematic outputs. Please if someone can review these and give a brief on these and then specify which output can be further improved.

@ajshajib

**Comment by ajshajib** [comment]
Pinging a few participants who have worked with MCMC, if any of you could provide help. 🙏 @JobairHasan @rafee-adnan @numjahan @Rajit13

**Comment by rafee-adnan** [comment]
@Nishijannat I can not open any of the link that you have provided. Kindly check if the notebooks are restricted or not.

**Reply by Rajit13** [reply]
Same with me!

**Reply by Nishijannat** [reply]
@rafee-adnan @Rajit13  extremely sorry. Fixed.

**Comment by ZareefJafar** [comment, marked as answer]
Although I am also working on mcmc. I can see two picks in some corner plots. Did you try increasing the `n_step` like 200,400,500 with lower fixed `n_walkers` like 100,200 etc for every run?
After that you can also try applying some `burnin` in the corner plot. `Burnin` value can be determined seeing when the trace plot of all the parameters  starts to converge. If some sharp increase occurs (`Too few points to create valid contours`) you can try applying some `thin` to reduce the density of plot.

**Reply by Nishijannat** [reply]
Bhaia I tried several times with different combinations of n_steps & walkers as you guys suggested, but its being just worse than before
Tried n_walker value 100 to 200 with n_steps range from 500 to 3000

**Reply by rafee-adnan** [reply]
Kindly read my comment @Nishijannat. You don't have to experiment with the `n_walker` value. Keep it fixed to `100`. You have to keep running the MCMC until it converges. Can you share your recent notebooks please?

**Reply by ZareefJafar** [reply]
Just noticed that you did not pass the `mcmc_backend = "mcmc_backend.h5"` to save the state after every iteration. So, I think you are starting from the beginning with every mcmc run.

Moreover as @rafee-adnan said, you need to keep  `n_walker` fixed with different or the same `n_step` in a series of runs of a particular trail. Make sure to  pass the same  `mcmc_backend` in `fitting_kwargs_list_mcmc` to save the previous states in the save file.

`Optional: `Remove the `"threadCount": 3` from kwargs to use full computational resources if you have no issues  with it. It will hopefully reduce the fitting time.

**Reply by Nishijannat** [reply]
@rafee-adnan  @ZareefJafar  actually  I tried with fixed walker value for few times and only then changed the value (when not improving). But yes, missed to use the backend file.  Now fixing this.

**Comment by Rajit13** [comment]
At first glance let me tell you my intuition without seeing any data value just the traceplot: Your traceplot seems off. How? It looks too smooth. 
![Untitled](https://github.com/ajshajib/BDLensing/assets/81336262/30317e72-a11b-4e08-b492-b9c6032ab251)
Where for your model I previously did this [Your PSO for attempt 3 and 4 run is far better than mine for which I found this traceplot]
![Untitled](https://github.com/ajshajib/BDLensing/assets/81336262/23340d77-3927-4cd3-80f9-478689ed227f)

**Comment by Rajit13** [comment]
Now to move on to the codes. You increased `n_walkers` instead of  `n_steps`/`n_runs`. You need to increase the steps and see after how many steps the blue part of the trace plot converges with the red line which is the median. My suggestions: use `n_walkers=80` and try the first 1000 steps. Then use `backend` to reiterate if needed. 

Then after you see that at certain steps it starts to converge use `burnin` to cut off useless parts of the traceplot and then use `thin` to plot corner plot from remaining [wanted] values after certain intervals.

**Comment by rafee-adnan** [comment]
I agree with @Rajit13 and @ZareefJafar. Kindly decrease the number of the walkers, `n_walkers` and increase the number of steps, `n_steps`. Suppose in this case you are working with 4000 walkers, and they only ran for 200 steps [[notebook 2]](https://colab.research.google.com/drive/1AKaUfEchmcaT7KVbEBLGYW-PnLgyVK00?usp=sharing). No matter how many walkers you take, if they don't get enough steps to converge to certain values (your model values), the MCMC would be of no use. So, taking higher number of steps is always advised until you reach a convergence. But you can keep the number of walkers fixed throughout your future MCMC runs. Like @Rajit13 suggested `n_walkers=80`, I worked with `n_walkers=100`. You can choose something around these numbers and re-run the MCMC. But don't forget to use `backend.h5` file. It is crucial for storing your previous steps and continuing the next run from there as it may require you to run this for ~5000 steps or more.

**Reply by Nishijannat** [reply]
Okay bhaia. Trying all the suggestions.

**Reply by Nishijannat** [reply]
@ZareefJafar @rafee-adnan @Rajit13 This is the latest [attempt](https://colab.research.google.com/drive/1c10Us2Vj23rC4l-10Ez-Y3d6PE77Meqv) . No matter how iterations are made, some values are not converging properly and also the convergence for other values not being smooth. I am currently running with more steps, meanwhile please check this one.

**Reply by ZareefJafar** [reply]
I think the convergence and corner plot looks great right now. Please share the plots.
If you want to improve further try increasing the lower, upper and initial value of source `beta`/ `beta_source_light1 `(low: 0.0997, high: 0.1008, init: 0.10) and source `R_sersic`/ `R_sersic_source_light0` (low: 0.05, high: 0.15, init: 0.0979). If this changes create any major problem in the convergence of other parameters do not need to do this. Everything looks great as is.

**Reply by rafee-adnan** [reply]
Like @ZareefJafar suggested increase the upper limit for both `R_sersic_source_light0` and `beta_source_light1` and re-run the mcmc. Other values converged pretty well. Additionally, I would recommend you to use these values of current mcmc run as initial value for a fresh mcmc run. I think it will also solve the issue of `e2_source_light0`, as it is creating a kind of bi-modal profile. Best of luck!

**Reply by rafee-adnan** [reply]
And you should not keep `gamma_lens` fixed for the mcmc run. When you start again your re-run, kindly make sure your gamma is not fixed. It should converge around `2.0`. @Nishijannat
